flink-clients/src/main/java/org/apache/flink/client/deployment/application/DetachedApplicationRunner.java

```diff
@@ -0,0 +1,82 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.client.deployment.application;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.client.ClientUtils;
+import org.apache.flink.client.deployment.application.executors.EmbeddedExecutor;
+import org.apache.flink.client.deployment.application.executors.EmbeddedExecutorServiceLoader;
+import org.apache.flink.client.program.PackagedProgram;
+import org.apache.flink.client.program.ProgramInvocationException;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.DeploymentOptions;
+import org.apache.flink.core.execution.PipelineExecutorServiceLoader;
+import org.apache.flink.runtime.dispatcher.DispatcherGateway;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletionException;
+
+import static java.util.Objects.requireNonNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * An {@link ApplicationRunner} which runs the user specified application
+ * using the {@link EmbeddedExecutor}.
+ * For this to work, this runner has to run on the same machine as the {@link DispatcherGateway}
+ * of the cluster, thus it does not go through REST to communicate with it.
+ *
+ * <p>In addition, this runner does not wait for the application to finish, but it submits the
+ * application in a {@code DETACHED} mode. As a consequence, applications with jobs that rely on
+ * operations like {@code [collect, print, printToErr, count]} will fail.
+ */
+@Internal
+public class DetachedApplicationRunner implements ApplicationRunner {
+
+	private static final Logger LOG = LoggerFactory.getLogger(DetachedApplicationRunner.class);
+
+	@Override
+	public List<JobID> run(final DispatcherGateway dispatcherGateway, final PackagedProgram program, final Configuration configuration) {
+		requireNonNull(dispatcherGateway);
+		requireNonNull(program);
+		requireNonNull(configuration);
+		return tryExecuteJobs(dispatcherGateway, program, configuration);
+	}
+
+	private List<JobID> tryExecuteJobs(final DispatcherGateway dispatcherGateway, final PackagedProgram program, final Configuration configuration) {
+		checkState(!configuration.get(DeploymentOptions.ATTACHED));
+
+		final List<JobID> applicationJobIds = new ArrayList<>();
+		final PipelineExecutorServiceLoader executorServiceLoader =
+				new EmbeddedExecutorServiceLoader(applicationJobIds, dispatcherGateway);
+
+		try {
+			ClientUtils.executeProgram(executorServiceLoader, configuration, program);
+		} catch (ProgramInvocationException e) {
+			LOG.warn("Could not execute application: ", e);
+			throw new CompletionException("Could not execute application.", e);
```

Comment:
Use `FlinkRuntimeException` or handle in its `supplyAsync` caller site. We don't tight this method to run in future context. do we?

Reply:
Maybe in a PR that I will open in the future, related to FLIP-85, but it is true that it is not needed here.